### PR TITLE
2nd Assignment by 202401885 (Soeun Kim)

### PR DESCRIPTION
**Pull request by soeun0609**
# Assignment Submission

## Student Information
- **Student ID**: [Your Student ID]
- **Name**: [Your Full Name]
- **Assignment**: [Assignment Number]

## Changed Files

```
submissions
└── 202401885
    └── assignment2
        ├── grader.py
        ├── problem1
        │   ├── README.md
        │   └── main.py
        ├── problem2
        │   ├── README.md
        │   └── textops
        │       ├── __init__.py
        │       ├── clean
        │       │   ├── __init__.py
        │       │   └── filters.py
        │       ├── main.py
        │       └── tokenize
        │           ├── __init__.py
        │           └── word.py
        ├── problem3
        │   ├── README.md
        │   └── main.py
        ├── problem4
        │   ├── README.md
        │   └── dsops
        │       ├── __init__.py
        │       ├── main.py
        │       ├── split
        │       │   ├── __init__.py
        │       │   └── train_test.py
        │       └── stats
        │           ├── __init__.py
        │           └── labels.py
        ├── problem5
        │   ├── README.md
        │   └── cachekit
        │       ├── __init__.py
        │       └── main.py
        └── problem6
            ├── README.md
            └── main.py
```

### Problem 1
- [ ] Implemented required functionality
- [ ] All tests pass locally

### Problem 2
- [ ] Implemented required functionality
- [ ] All tests pass locally

### Problem 3
- [ ] Implemented required functionality
- [ ] All tests pass locally

### Problem 4 (Optional)
- [ ] Implemented required functionality
- [ ] All tests pass locally

### Problem 5 (Optional)
- [ ] Implemented required functionality
- [ ] All tests pass locally

### Problem 6 (Optional)
- [ ] Implemented required functionality
- [ ] All tests pass locally

## Testing
- [ ] I have run `python grader.py` locally
- [ ] All implemented problems pass their tests
- [ ] I have tested edge cases mentioned in the problem statements

## Additional Notes

---
**Note**: This PR will be automatically validated for:
- Correct directory structure
- Valid Python syntax
- PR title format: `Nth Assignment by {student_id} (Full Name)`
- File organization compliance

### CODE REVIEW

File: submissions/202401885/assignment2/problem2/textops/tokenize/word.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    """
+    규칙:
+    - 단일 공백 기준 분할
+    - 빈 문자열/공백뿐이면 [] 반환
+    - 입력이 비정규화여도 안전하게 동작하도록 내부적으로 공백 축약 후 split
+    """
+    if not isinstance(s, str):
+        raise TypeError("word_tokens: input must be str")
+
+    s = re.sub(r"\s+", " ", s).strip()
```
Comment (hist0613) [inline submissions/202401885/assignment2/problem2/textops/tokenize/word.py:14]: 이게 무슨 작업을 하는 코드인지 comment 를 좀 남겨주실래요?

**Review by hist0613 — CHANGES_REQUESTED** [review]
problem1 외에는 나머지 모두 다 `grader.py` 를 통과했습니다.
problem1 디렉토리 구조 확인해주세요~

File: submissions/202401885/assignment2/problem2/textops/clean/filters.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    # 1) 소문자화
+    s = s.lower()
+
+    # 2) ASCII punctuation 중 '와 -만 남기고 나머지 제거
+    # string.punctuation: !"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~
+    punct_to_remove = string.punctuation.replace("'", "").replace("-", "")
+    s = s.translate(str.maketrans("", "", punct_to_remove))
+
+    # 3) 모든 공백 축약 + 트림
+    s = re.sub(r"\s+", " ", s).strip()
```
Comment (hist0613) [inline submissions/202401885/assignment2/problem2/textops/clean/filters.py:24]: 흐음 👀

File: submissions/202401885/assignment2/problem2/textops/tokenize/word.py (hunk 1 above)
Comment (hist0613) [inline submissions/202401885/assignment2/problem2/textops/tokenize/word.py:14]: `word_tokens` 에 어떤 입력이 들어오는지 문제 정의를 다시 확인해주세요!

File: submissions/202401885/assignment2/problem3/main.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
+"""
+Problem 3 — tokenstats (module + __main__ demo)
+- count token frequencies
+- get top-k (freq desc, token asc for ties)
+"""
+
+from typing import List, Dict, Tuple
+
+def count_tokens(tokens: List[str]) -> Dict[str, int]:
```
Comment (hist0613) [inline submissions/202401885/assignment2/problem3/main.py:9]: 오호오... 👀

File: submissions/202401885/assignment2/problem1/main.py
[the remark this answers is not in the document]
Comment (soeun0609) [reply submissions/202401885/assignment2/problem1/main.py:1]: 수정했습니다! 알려주셔서 감사합니다ㅎㅎ

**Review by hist0613 — CHANGES_REQUESTED** [review]
credit 적절하게 해주세요 :)

File: submissions/202401885/assignment2/problem1/problem1.py
[the remark this answers is not in the document]
Comment (soeun0609) [reply submissions/202401885/assignment2/problem1/problem1.py:1]: 수정했습니다!
